File: src/de/grzb/Operator.java
package de.grzb;

import java.util.Arrays;
import java.util.List;

class Operator implements IChainable {
    private static Operator ADD        = new Operator("+", 2);
    private static Operator SUBTRACT   = new Operator("-", 2);
    private static Operator MULTIPLY   = new Operator("*", 1);
    private static Operator DIVIDE     = new Operator("/", 1);
    private static Operator POTENTIATE = new Operator("^", 0);

    private final String name;
    private final int precedence;
    private static List<Operator> values;

    static {
        Operator.values = Arrays.asList(ADD, SUBTRACT, MULTIPLY, DIVIDE, POTENTIATE);
    }

    private Operator(String name, int precedence) {
        this.name = name;
        this.precedence = precedence;
    }

    public int getPrecedence() {
        return this.precedence;
    }

    public static List<Operator> values() {
        return Operator.values;
    }

    public static Operator getByName(String name) {
        for(Operator value : Operator.values()) {
            if(value.name.equals(name)) {
                return value;
            }
        }

        return null;
    }

    public Number calc(Number operand1, Number operand2) {
        if(this.equals(ADD))             return Operator.add(operand1, operand2);
        else if(this.equals(SUBTRACT))   return Operator.subtract(operand1, operand2);
        else if(this.equals(MULTIPLY))   return Operator.multiply(operand1, operand2);
        else if(this.equals(DIVIDE))     return Operator.divide(operand1, operand2);
        else /* if(this.equals(POTENTIATE))*/ return Operator.potentiate(operand1, operand2);
    }

    public static Number add(Number operand1, Number operand2) {
        return new Number(operand1.getValue() + operand2.getValue());
    }

    public static Number subtract(Number operand1, Number operand2) {
        return new Number(operand1.getValue() - operand2.getValue());
    }

    public static Number multiply(Number operand1, Number operand2) {
        return new Number(operand1.getValue() * operand2.getValue());
    }

    public static Number divide(Number operand1, Number operand2) {
        return new Number(operand1.getValue() / operand2.getValue());
    }
    public static Number potentiate(Number operand1, Number operand2) {
        return new Number(Math.pow(operand1.getValue(), operand2.getValue()));
    }

    public String toString() {
        return this.name;
    }

    @Override
    public IChainable copy() {
        return this;
    }

}

File: src/operator.cpp
#include <operator.hpp>
#include <util.hpp>
#include <vector>

using namespace util;

std::vector<Glib::ustring> Operator::operators = {
    "+",
    "-",
    "*",
    "/",
    "^"
};

Operator::Operator(const int op_type, const int precedence) {
    this->op_type     = op_type;
    this->precedence = precedence;
}
Operator::~Operator() {}
Operator* Operator::get(const int op_type) {
    switch(op_type) {
        case ADD:
            return new Operator(op_type, 0);
        case SUBTRACT:
            return new Operator(op_type, 0);
        case MULTIPLY:
            return new Operator(op_type, 1);
        case DEVIDE:
            return new Operator(op_type, 1);
        case POTENTIATE:
            return new Operator(op_type, 2);
        default:
            throw std::invalid_argument(str(op_type) + " is not a valid operator enum!");
    }
}
Operator* Operator::get(const Glib::ustring op_symbol) {
    if(op_symbol == "+") {
        return Operator::get(ADD);
    }
    else if(op_symbol == "-") {
        return Operator::get(SUBTRACT);
    }
    else if(op_symbol == "*") {
        return Operator::get(MULTIPLY);
    }
    else if(op_symbol == "/") {
        return Operator::get(DEVIDE);
    }
    else if(op_symbol == "^") {
        return Operator::get(POTENTIATE);
    }

    throw std::invalid_argument(op_symbol + " is not a valid operator symbol!");
}
long double Operator::add(const long double u, const long double v) {
    return u + v;
}
long double Operator::subtract(const long double u, const long double v) {
    return u - v;
}
long double Operator::multiply(const long double u, const long double v) {
    return u * v;
}
long double Operator::devide(const long double u, const long double v) {
    return u / v;
}
long double Operator::potentiate(const long double u, const long double v) {
    return std::pow(u, v);
}
long double Operator::calc(const long double u, const long double v) {
    switch(this->op_type) {
        case ADD:
            return Operator::add(u, v);
        case SUBTRACT:
            return Operator::subtract(u, v);
        case MULTIPLY:
            return Operator::multiply(u, v);
        case DEVIDE:
            return Operator::devide(u, v);
        case POTENTIATE:
            return Operator::potentiate(u, v);
        default:
            throw std::invalid_argument(str(op_type) + " is not a valid operator enum!");
    }
}
Glib::ustring Operator::to_string() {
    return this->operators[this->op_type];
}
bool Operator::is_operator(const Glib::ustring op) {
    if(op == "+" || op == "-" || op == "*" || op == "/" || op == "^") {
        return true;
    }
    return false;
}
